boards: shields: add suport for weact_ov2640_cam_module

CharlesDias opened the pull request:
Add support for OV2640 camera sensor to use with WeAct Studio MiniSTM32H7xx boards.

PR dependency:
- https://github.com/zephyrproject-rtos/zephyr/pull/71463

CharlesDias committed:
boards: shields: add suport for weact_ov2640_cam_module

Add support for WeAct Studio MiniSTM32H7xx OV2640 camera sensor.

Signed-off-by: Charles Dias <charlesdias.cd@outlook.com>

CharlesDias commented:
> I think we should define a binding for the connector (similar to `dsi_lcd_qsh_030` display connector for example) since it's apparently used by many vendors ([#71463 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/71463#discussion_r1713617227)) and use that binding to specify the pins in ov2640 node, in a board-independent way.
> 
> Also, the I2C, DCMI (except `port` child node, and board-independent properties like `sensor`), and DMA nodes should be moved to a board-specific overlay (#77854 does the same for display shield) & then be referenced by generic aliases (like "cam_interface" for DCMI so the sensor can be used with a board that has a cam interface other than DCMI).
> 
> BTW, should the shield be renamed to "ov2640_cam_module" for example?

Thanks for your review, @JarmouniA! 

Since I'm not familiar with the connector binding approach, I'll take a look at it carefully.

CharlesDias commented:
Hi, @JarmouniA.

> I think we should define a binding for the connector (similar to `dsi_lcd_qsh_030` display connector for example) since it's apparently used by many vendors ([#71463 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/71463#discussion_r1713617227)) and use that binding to specify the pins in ov2640 node, in a board-independent way.

I think it is done! Let me know if I got right your suggestion. 

> BTW, should the shield be renamed to "ov2640_cam_module" for example?

About the shield name, could it be "weact_ov2640_cam_module"? To make it less generic.

CharlesDias commented:
> My point was that the connector should't be considered a WeAct one, it's used on a lot of boards from other vendors as well (like ESP32-CAM https://components101.com/sites/default/files/inline-images/ESP32-CAM-Schematics.png), so if we are going to make a compatible for it, it should not have WeAct in its name. Same thing for the camera module (https://components101.com/sites/default/files/components/ESP32-CAM-Camera-Module.jpg).

Good observation, @JarmouniA. 

There is one point here to consider: even though the connector is similar, the position of the pins is different from the WeAct schematic https://github.com/WeActStudio/MiniSTM32H7xx/blob/master/HDK/STM32H7xx%20SchDoc%20V11.pdf. I believe this is a problem when creating a generic OV2640 shield version. What you think?

CharlesDias commented:
I placed the connectors next to each other and noticed that the pins are upside down. Additionally, the ESP32-CAM doesn't connect the PWDN pin, while the WeAct doesn't connect the RESET pin (which is linked to the reset system).

![image](https://github.com/user-attachments/assets/a3d72a76-be68-4d77-b74e-912bd1e58132)

CharlesDias commented:
> > the position of the pins is different from the WeAct schematic
> 
> Ah, good catch, didn't pay attention to this.
> 
> In this case, would it make sense to have a generic compatible for the connector, then specific ones for each vendor, depending on the pins mapping, that include the common one? Or would it be simpler to just keep the current implementation? The same for the cam module.

I think we would keep it simpler as in the current implementation. Is it ok for you?

CharlesDias commented:
Thank you @JarmouniA for your valuable review!

CharlesDias commented:
Change the shield name to `weact_ov2640_cam_module`.

josuah commented:
This introduces a new naming for shields.
As noted here, there seems to be a `zephyr_` prefix for all shields, which might allow to make a difference between names specific to a SoC/Boad/project and those standardized by Zephyr, facilitating review of all standards:
- https://github.com/zephyrproject-rtos/zephyr/pull/77854#discussion_r1740830325

There is a compromise to take in that:

- `dvp` is relatively widespread but roots back to OmniVision
- `dcmi` is a peripheral by ST
- `camera_interface` sounds like the [right name](https://en.wikipedia.org/wiki/Camera_interface) but could be mistaken with MIPI-CSI which also "a camera interface".
- `camif` does not seems as broadly as the vendor terms.

If we try to copy-cat `zephyr_mipi_dsi`, this could be:

- `zephyr_camif` or `zephyr_dvp` or `zephyr_camera_dvp` or zephyr_camera_parallel`
- `zephyr_camera_i2c` or `zephyr_camif_i2c` or `zephyr_dvp_i2c` or `zephyr_camera_parallel_i2c` or `zephyr_mipi_csi_i2c`

Which one sounds the best?

CharlesDias commented:
> Which one sounds the best?

Hi, @josuah. The `zephyr_camera_dvp` and `zephyr_camera_i2c`sounds great for me!
Thanks for your suggestion!

CharlesDias commented:
- Change `camera_interface` to `zephyr_camera_dvp`.
- Change `camera_connector_i2c` to `zephyr_camera_i2c`.

josuah reviewed (dismissed):
The `remote-endpoint` is present but not actively used yet as not supported by Zephyr devicetree syntax. This should not be any problem to be sitting there AFAIU from #57708.

@JarmouniA I tried to add your other account but seems inactive, it is still possible to make reviews even without collaborator access if you wish.

josuah commented:
CI complains about `samples/drivers/video/capture_to_lvgl/` not existing, which will be introduced by #71463.

erwango commented:
@CharlesDias There are some CI issues to be fixed

CharlesDias commented:
Hi @erwango. 
This PR https://github.com/zephyrproject-rtos/zephyr/pull/71463 has to be merged first. Could you review it? Thank you!

CharlesDias commented:
Rebase.

CharlesDias commented:
Add new MCO configuration as suggested on https://github.com/zephyrproject-rtos/zephyr/pull/71463#discussion_r1773212375

CharlesDias commented:
Rebase.

josuah commented on boards/shields/weact_ov2640_cam_module/Kconfig.shield:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config SHIELD_WEACT_OV2640_CAM_MODULE
+	def_bool $(shields_list_contains,weact_ov2640_cam_module)
```
I did not know Zephyr had custom [Kconfig](https://docs.zephyrproject.org/latest/build/kconfig/preprocessor-functions.html) macros!

CharlesDias replied:
Zephyr is full of surprises! :)

josuah commented on boards/shields/weact_ov2640_cam_module/doc/index.rst:
```diff
@@ -0,0 +1,82 @@
+Requirements
+************
+
+Your board needs to have a ``dcmi`` device tree label to work with this shield.
```
Is it now `zephyr_camera_dvp`?
https://github.com/zephyrproject-rtos/zephyr/compare/d73a2d869c5c426f64aea6860b63205a42a1fbbf..76759c81c076ea47aebe244b2f5daa940ae32eca#diff-a04707eebe85f0f2eef1831b4061cffe823f57bbf4a2ac8f549d08a167f3e106L28

Maybe it was to mean that it is only expected to work with DCMI in the current state... in which case "devicetree label".

Also, `device tree` -> `devicetree` according to this search, also also used interchangeably it seems...
https://docs.zephyrproject.org/latest/gsearch.html?q=device+tree&check_keywords=yes&area=default#gsc.tab=0&gsc.q=device%20tree&gsc.page=1

josuah reviewed (changes requested):
Some very minor documentation edits spotted but everything else looks ready to me!

CharlesDias commented:
Update the [samples/drivers/video/capture_to_lvgl/sample.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/bda49ac20bcd88c2381ecdc19894630d5b7d4649/samples/drivers/video/capture_to_lvgl/sample.yaml) file to include this shield.

CharlesDias commented:
Hello, everyone.

I am trying to understand why the CI is failing! Do you have any suggestions for me? 

Another point: should I update the [samples/drivers/video/capture_to_lvgl/sample.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/bda49ac20bcd88c2381ecdc19894630d5b7d4649/samples/drivers/video/capture_to_lvgl/sample.yaml) to include this shield? If that's the case, should it be like this?

```
sample:
  name: Video capture to LVGL
tests:
  sample.video.capture_to_lvgl:
    tags:
      - video
      - shield
      - samples
    extra_args:
      - platform:mini_stm32h743:SHIELD=weact_ov2640_cam_module
    harness: console
    harness_config:
      fixture: fixture_camera
      type: multi_line
      ordered: true
      regex:
        - "Device name"
        - "Format"
        - "Capture started"
    platform_allow:
      - mini_stm32h743
    depends_on: video
    integration_platforms:
      - mini_stm32h743
```

CharlesDias commented:
> @CharlesDias
> 
> ```
> extra_args:
>       - platform:mini_stm32h743:SHIELD=weact_ov2640_cam_module
> ```
> 
> this does not look right.
> 
> See
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/3a15e37b5a9a0d17c111eecb3a1803c2ab20cd5d/samples/drivers/display/sample.yaml#L136

Thanks, @JarmouniA! Do I have to change also from `sample.video.capture_to_lvgl:` to `sample.video.capture_to_lvgl.shield.weact_ov2640_cam_module:`?

CharlesDias commented:
I used as a model the https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/drivers/video/capture/sample.yaml file.

CharlesDias commented:
Not yet, @JarmouniA! There is something more! Twister is mysterious to me! :)

josuah commented on boards/shields/weact_ov2640_cam_module/Kconfig.defconfig:
```diff
@@ -0,0 +1,12 @@
+
+if SHIELD_WEACT_OV2640_CAM_MODULE
+
+config VIDEO_HFLIP
+	default n
+
+config VIDEO_VFLIP
+	default n
```
I think this is where Twister complains now: it needs something like `bool Description of this`.

josuah replied:
```
warning: VIDEO_HFLIP (defined at boards/shields/weact_ov2640_cam_module/Kconfig.defconfig:6) defined without a type

warning: VIDEO_VFLIP (defined at boards/shields/weact_ov2640_cam_module/Kconfig.defconfig:9) defined without a type

error: Aborting due to Kconfig warnings
```

CharlesDias commented:
Removed the [boards/shields/weact_ov2640_cam_module/Kconfig.defconfig](https://github.com/zephyrproject-rtos/zephyr/compare/5cf0686b9c4423d8e672033fb934cca6c18bac6d..54760b90635dc46f16a250fc298fe638fc2f5776#diff-ea21e0c2c49182c2419f4292fa61b0efb8bf48cd8593319fda544d69736d800d) file.

CharlesDias commented:
Rebase

josuah commented on boards/shields/weact_ov2640_cam_module/weact_ov2640_cam_module.overlay:
```diff
@@ -0,0 +1,44 @@
+	capture-rate = <1>;
+
+	port {
+		dcmi_ep_in: endpoint {
```
Maybe this would need an alias to be as generic as the `zephyr_camera_dvp`? `zephyr_camera_dvp_ep_in`? Or less verbose `zephyr_camera_dvp_in`? Or something else?

josuah commented on boards/shields/weact_ov2640_cam_module/weact_ov2640_cam_module.overlay:
```diff
@@ -0,0 +1,44 @@
+
+		port {
+			ov2640_ep_out: endpoint {
+				remote-endpoint = <&dcmi_ep_in>;
```
See comment below about what to pick to make it generic.

josuah reviewed (dismissed):
It looks in good condition to be merged.
The last little details that might have been missed will be a lot easier to fix once an alternative camera module is to be supported for this board, or once an alternative board tries to use that camera module.

Thank you for this integration!

CharlesDias commented:
Hi, everyone.

What you think about the use of `CONFIG_VIDEO_HFLIP=y` and `CONFIG_VIDEO_VFLIP=y`symbols? In this way, the shield depends on these symbols that are relevant only for sample/drivers/video/capture_to_lvgl code. Does this make sense? Is there another way? Thanks!

CharlesDias commented:
Change port label from `dcmi_ep_in` to `zephyr_camera_dvp_in`.

josuah commented:
> What you think about the use of `CONFIG_VIDEO_HFLIP=y` and `CONFIG_VIDEO_VFLIP=y` symbols inside of [boards/shields/weact_ov2640_cam_module/boards/mini_stm32h743.conf](https://github.com/zephyrproject-rtos/zephyr/pull/77949/files#diff-8e3a95d81dbc55b3f05d19e162b50a7488bd57e9ffb66b9e97d214400118e05a)? In this way, the shield depends on these symbols that are relevant only for sample/drivers/video/capture_to_lvgl code. Does this make sense? Is there another way? Thanks!

The FPC can eventually be folded to rotate the camera by 180°, but then that can still be overridden by the user.
This reminds me of [`LIBV4LCONTROL_FLAGS`](https://github.com/philips/libv4l/blob/master/libv4lconvert/control/libv4lcontrol.c#L369-L371) that permits to [rotate a misoriented camera](https://techieboycdo.blogspot.com/2010/08/fix-upside-down-or-inverted-webcam-on.html) at the very last moment.

kartben commented on boards/shields/weact_ov2640_cam_module/boards/mini_stm32h743.overlay:
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) Copyright (c) 2024 Charles Dias
```
typo

kartben commented on boards/shields/weact_ov2640_cam_module/weact_ov2640_cam_module.overlay:
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) Copyright (c) 2024 Charles Dias
```
typo

kartben commented on samples/drivers/video/capture_to_lvgl/Kconfig:
can you put this in a different commit please?

kartben reviewed (changes requested):
one small comment, looks good otherwise!

CharlesDias committed:
samples: drivers: video: capture_to_lvgl: add menu and update the YAML

Add the Kconfig menu to the sample configuration.
Update sample.yaml by adding the shield configuration.

Signed-off-by: Charles Dias <charlesdias.cd@outlook.com>
